Update "Data fixture annotation" documentation (#9273)
magento/devdocs: src/guides/v2.4/test/integration/annotations/magento-data-fixture.md

---

A data fixture is a PHP script that sets data you want to reuse in your test.

> The script can be defined in a separate PHP file, class or as a local test case method.

Reviewer: ```suggestion
The script can be defined in a separate PHP file, class, or as a local test case method.
```

Use data fixtures to prepare a database for tests.
The Integration Testing Framework (ITF) reverts the database to its initial state automatically.
To set up a date fixture, use the `@magentoDataFixture` annotation.

## Format

> `@magentoDataFixture` takes an argument that points to the data fixture as a filename, full class name or local method.

Reviewer: ```suggestion
`@magentoDataFixture` takes an argument that points to the data fixture as a filename, full class name, or local method.
```

1. Fixtures declared at a test level have a higher priority than fixtures declared at a test case level.
1. Test case fixtures are applied to each test in the test case, unless a test has its own fixtures declared.
1. Annotation declaration at a test case level does not affect tests that have their own annotation declarations.
1. Fixture alias SHOULD be camelcase.
1. Fixture data provider MUST be a valid JSON string.

## Usage

As mentioned above, there are three ways to declare fixtures:

-  as a PHP script file that is used by other tests and test cases.
-  as a local method that is used by other tests in the test cases.

> -  as a [Class][parameterizedDataFixture] that implements `Magento\TestFramework\Fixture\DataFixtureInterface` or `Magento\TestFramework\Fixture\RevertibleDataFixtureInterface`

Reviewer: ```suggestion
-  as a [Class][parameterizedDataFixture] that implements `Magento\TestFramework\Fixture\DataFixtureInterface` or `Magento\TestFramework\Fixture\RevertibleDataFixtureInterface`.
```

1. Test case fixtures are applied to each test in the test case, unless a test has its own fixtures declared.
1. Annotation declaration at a test case level does not affect tests that have their own annotation declarations.

### For Parameterized Data Fixtures

{:.bs-callout-info}
Parameterized Data Fixtures currently only available for Magento Open Source contributors, and will be released for public with Magento Open Source 2.5.5.

> 

Author: It's an accident leftover from when I misunderstood you and moved the principles to data fixture provider, but now according to your note above I am bringing two of them back here.

1. Test case fixtures are applied to each test in the test case, unless a test has its own fixtures declared.
1. Annotation declaration at a test case level does not affect tests that have their own annotation declarations.

> ### For Parametrized Data Fixtures

Reviewer 2: Insert blank line after line 37

1. Fixture class MUST implement `Magento\TestFramework\FixtureInterface`
1. Fixture class MUST implement `Magento\TestFramework\RevertibleFixtureInterface` if the data created by the fixture

>    is revertible. For instance fixture that creates an entity (e.g. product).

Reviewer 2: ```suggestion
   is revertible. For instance a fixture that creates an entity (e.g. product).
```

1. Fixture class namespace MUST match `Magento\<Module Name>\Fixture`

> 1. Fixture class SHOULD follow single responsibility principle. See example with decoupled fixtures class (link?).

Reviewer 2: There should be a link where it says (link?)

[`dev/tests/integration/testsuite/Magento/Cms/Controller/PageTest.php`][] demonstrates an example of the `testCreatePageWithSameModuleName()` test method that uses data from the `cmsPageWithSystemRouteFixture()` data fixture.

### Data Fixture Data Provider

{:.bs-callout-info}

> Data Fixture Data Provider is only applicable to Parameterized Data Fixture and is currently only available for Magento Open Source contributors, and will be released for public with Magento Open Source 2.4.5.

Reviewer: ```suggestion
Data Fixture Data Provider is only applicable to Parameterized Data Fixture and is currently only available for Magento Open Source contributors. It will be released for general use with Magento Open Source 2.4.5.
```

There are two types of data providers:

-  Inline JSON as data provider
-  [`@magentoDataFixtureDataProvider`][magentoDataFixtureDataProvider] annotation

#### Inline JSON as data provider

> Data can be passed to the [Parameterized Data Fixture][parameterizedDataFixture] using `with` directive as follows:

Reviewer: ```suggestion
Data can be passed to the [Parameterized Data Fixture][parameterizedDataFixture] using the `with` directive as follows:
```

Example:

```php?start_inline=1
class ProductsList extends \PHPUnit\Framework\TestCase
{
    /**
     * @magentoDataFixture Magento\Catalog\Test\Fixture\Product with:{"price": 5.0}
     * @magentoDataFixture Magento\Catalog\Test\Fixture\Product with:{"price": 10.0}
     * @magentoDataFixture Magento\Catalog\Test\Fixture\Product with:{"price": 15.0}
     */
    public function testGetProductsCount(): void
    {
    }
}
```

### Fixture Alias

{:.bs-callout-info}

> Fixture Alias is only applicable to Parameterized Data Fixture and is currently only available for Magento Open Source contributors, and will be released for public with Magento Open Source 2.4.5.

Reviewer: ```suggestion
Fixture Alias is only applicable to Parameterized Data Fixture and is currently only available for Magento Open Source contributors. It will be released for general use with Magento Open Source 2.4.5.
```

> [Parameterized Data Fixture][parameterizedDataFixture] can be given an alias using directive `as`. The fixture alias is used as reference to retrieve the data returned by the fixture and also as reference in other fixtures parameters.

Reviewer: ```suggestion
You can give [Parameterized Data Fixture][parameterizedDataFixture] an alias using the `as` directive. The fixture alias is used as a reference to retrieve the data returned by the fixture and also as a reference in other fixtures parameters.
```

#### Retrieve fixture data in the test

A test can retrieve data that was returned by a [Parameterized Data Fixture][parameterizedDataFixture] using `Magento\TestFramework\Fixture\DataFixtureStorageManager` and the fixture alias.

The following example shows how to retrieve data that was returned by the fixtures:

```php?start_inline=1
class ProductsList extends \PHPUnit\Framework\TestCase
{
    /**
     * @magentoDataFixture Magento\Catalog\Test\Fixture\Product as:product1
     * @magentoDataFixture Magento\Catalog\Test\Fixture\Product as:product2
     * @magentoDataFixture Magento\Catalog\Test\Fixture\Product as:product3
     */
    public function testGetProductsCount(): void
    {
        $fixtures = DataFixtureStorageManager::getStorage();
        $product1 = $fixtures->get('product1');
        $product2 = $fixtures->get('product2');
        $product3 = $fixtures->get('product3');
    }
}
```

#### Supply data to parameterized data fixture as a variable

It is possible to supply data as a variable from one fixture to another using the fixture alias in one of the following formats:

> -  `$fixtureAlias$` is a reference to the data that was returned by the fixture with alias `fixtureAlias`

Reviewer: ```suggestion
-  `$fixtureAlias$` is a reference to the data that was returned by the fixture with alias `fixtureAlias`.
```

[`dev/tests/integration/testsuite/Magento/Cms/Controller/PageTest.php`][] demonstrates an example of the `testCreatePageWithSameModuleName()` test method that uses data from the `cmsPageWithSystemRouteFixture()` data fixture.

### Parametrized Fixture class

Define the fixture in a separate Parametrized Fixture class when you think it could be reused in different test cases.
Fixture class can be provided with a data provider for additional parameters and customizations.

> There are two types of data providers:

Reviewer 2: Insert blank line after line 103

-  Inline JSON as data provider
-  Decoupled fixtures class

#### Inline JSON as data provider

Example:

```php?start_inline=1
class ProductsList extends \PHPUnit\Framework\TestCase
{
    /**

>     * @magentoDataFixture \Magento\Catalog\Fixture\CreateSimpleProduct with:{"sku": "simple1", "price": 5.0}
>     * @magentoDataFixture \Magento\Catalog\Fixture\CreateSimpleProduct with:{"sku": "simple2", "price": 10.0}
>     * @magentoDataFixture \Magento\Catalog\Fixture\CreateSimpleProduct with:{"sku": "simple3", "price": 15.0}

Reviewer 3: ```suggestion
    * @magentoDataFixture \Magento\Catalog\Test\Fixture\Product with:{"price": 5.0}
    * @magentoDataFixture \Magento\Catalog\Test\Fixture\Product with:{"price": 10.0}
    * @magentoDataFixture \Magento\Catalog\Test\Fixture\Product with:{"price": 15.0}
```

    */
    public function testGetProductsCount(): void
    {
    }
}
```

### A class method as data provider

Example:

```php?start_inline=1
class ProductsList extends \PHPUnit\Framework\TestCase
{
   /**
   * @magentoDataFixture Magento\Catalog\Fixture\CreateSimpleProduct as:product1
   * @magentoDataFixture Magento\Catalog\Fixture\CreateSimpleProduct as:product2
   * @magentoDataFixture Magento\Catalog\Fixture\CreateSimpleProduct as:product3
   */
   public function testGetProductsCount(): void
   {
   }

   public function getProductsCountFixtureDataProvider(): array
   {
      return [
         [
             'product1' => [
                 'sku' => 'simple1'
             ],
             'product2' => [
                 'sku' => 'simple2'
             ],
             'product3' => [
                 'sku' => 'simple3',
                 'status' => Status::STATUS_DISABLED,
             ],
         ]
      ];
   }
}
```

> 

Reviewer 3: We will have to move this section  to a new page under @magentoDataFixtureDataProvider

> ### Decoupled fixtures class

Reviewer 3: I suggest to move this section to a different page under "Best Practices For using Parameterized Fixtures"
